scalyr_agent/log_processing.py

```diff
@@ -1104,21 +1115,47 @@ def __refresh_pending_files(self, current_time):
                         # just mark this file portion as now invalid.
                         current_log_file.valid = False
+
                     current_log_file.is_log_file = False
                     current_log_file.position_end = (
                         current_log_file.position_start
                         + current_log_file.last_known_size
                     )
-                    # Note, we do not yet detect if current_log_file is actually pointing to the same inode as the
-                    # log_path.  This could be true if the log file was copied to another location and then truncated
-                    # in place (a commom mode of operation used by logrotate).  If this is the case, then the
-                    # file_handle in current_log_file will eventually fail since it will seek to a location no longer
-                    # in the file.  We handle that fairly cleanly in __fill_buffer so no need to do it here.  However,
-                    # if we want to look for the file where the previous log file was copied, this is where we would
-                    # do it.  That is a future feature.
-
-                    # Add in an entry for the file content at log_path.
-                    self.__add_entry_for_log_path(latest_inode)
+                    if (
+                        self.__file_system.trust_inodes
+                        and latest_inode == current_log_file.inode
+                        and self.__enable_copy_truncate_log_rotation
+                    ):
+                        # copy-truncate log rotation
+                        copied_file = self.__find_copy_truncate_file()
+                        if copied_file is not None:
+                            # Add copied log file before truncated log file
+                            self.__add_entry_for_log_path(
+                                None, copied_log_path=copied_file, index=-1
+                            )
+                            # Modify truncated log file to start at end of copied log file
+                            copied_log_file = self.__pending_files[-2]
+                            truncated_log_file = self.__pending_files[-1]
+                            truncated_log_file.position_start = (
+                                copied_log_file.position_end
+                            )
+                            truncated_log_file.position_end = (
+                                truncated_log_file.position_start + latest_size
+                            )
+                            truncated_log_file.last_known_size = latest_size
+                            truncated_log_file.is_log_file = True
+                            copied_log_file.is_log_file = False
+                        else:
+                            log.warning(
+                                "Could not find copied log file for copy-truncate log rotation file."
```

Comment:
Let's modify this warning to be "It appears that the file was rotated using copy-truncate but the copied file could not be found.  This may mean some log lines are missing."

This will be emitted every time the file is rotated... but I guess we think that won't be that often where this log message will amount to many bytes.  Otherwise I'd suggest we rate limit the line.